### ISSUE TITLE
(SEIZURE WARNING) Dynamic bone visualization can be seizure inducing when sweeping a laser over the inspector.

### ISSUE THREAD

**Issue by ModernBalloonie** [Yellow-Dog-Man/Resonite-Issues#2459]
### Describe the bug?

It seems like when a bone has specifically multiple child objects, and you visualze the bones while messing with the inspector, the debug visuals strobe really fast.

### To Reproduce

resrec:///U-ModernBalloonie/R-6c24cd3e-3462-42ca-a5a5-db8786894215

Here's a world I made to replicate the issue. Click visualize bones and over over the options in the inspector. See how they make the bone visuals freak out.

### Expected behavior

For the debug visuals to not strobe.

### Screenshots

SEIZURE WARNING!

https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/82356528/c2442a87-cdb1-4c82-83bc-1bbfc47d343e

### Resonite Version Number

Beta 2024.6.25.1149

### What Platforms does this occur on?

Windows

### What headset if any do you use?

Quest Pro

### Log Files

[DESKTOP-QBGCSAQ - 2024.6.25.1149 - 2024-07-02 20_09_20.log](https://github.com/user-attachments/files/16076544/DESKTOP-QBGCSAQ.-.2024.6.25.1149.-.2024-07-02.20_09_20.log)

### Additional Context

Sometimes I find it really hard to work due to this issue, and it gives me a headache and makes me dizzy.

### Reporters

modernballoonie

**Comment by shiftyscales** [comment]
This primarily seems to be an issue when the "simulate terminal bones" setting is enabled- with it not enabled, the strobing does not occur- although the debug visuals are a bit unstable while the object is held- even perfectly still while in desktop mode.

I suspect that the added debug visuals from the simulated terminal bones are overlapping and z-fighting causing the strobing observed.

Does the issue occur if there's only a single side-chain per bone in the primary chain, @ModernBalloonie?

I think part of the issue is the setup, and how many branching side-chains you have- in particular as well that all of them are only one bone long.

Could you also check if the issue occurs if the side chains are longer than just a single bone?

**Comment by Banane9** [comment]
> I suspect that the added debug visuals from the simulated terminal bones are overlapping and z-fighting causing the strobing observed.
> 
> Does the issue occur if there's only a single side-chain per bone in the primary chain, @ModernBalloonie?
> 
> I think part of the issue is the setup, and how many branching side-chains you have- in particular as well that all of them are only one bone long.

Ah, I was wondering why I've never seen anything like that happen, but that would seem like a very likely explanation

**Comment by ModernBalloonie** [comment]
> This primarily seems to be an issue when the "simulate terminal bones" setting is enabled- with it not enabled, the strobing does not occur- although the debug visuals are a bit unstable while the object is held- even perfectly still while in desktop mode.
> 
> I suspect that the added debug visuals from the simulated terminal bones are overlapping and z-fighting causing the strobing observed.
> 
> Does the issue occur if there's only a single side-chain per bone in the primary chain, @ModernBalloonie?
> 
> I think part of the issue is the setup, and how many branching side-chains you have- in particular as well that all of them are only one bone long.
> 
> Could you also check if the issue occurs if the side chains are longer than just a single bone?

It doesn't occur like this:
![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/82356528/337384b4-8239-44c5-a615-244a0266c775)

Nor this:
![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/82356528/dbfd6f78-acf8-48cc-b89f-5b603849e22f)

But if I add a few more boxes, instability starts.
What's REALLY interesting though is that regardless, even if I have a bunch of seperated bone chains... if I open an avatar in an inspector, and mouse over the options, look what happens. These are all seperate chains.

[2024-07-08 15-23-16-1.webm](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/82356528/a049392c-6ac5-42c4-99cb-bb5ee9716f82)

The visuals seem to be connected to eachother in some way.

**Comment by ModernBalloonie** [comment]
I couldn't get it to happen with one chain, but as you can see, I duplicate this chain, and it seems like the debug visuals are freaking out again.
[2024-07-08 15-34-06-1.webm](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/82356528/343b27d0-bf60-4566-b57a-2a782b9eeecb)

**Comment by shiftyscales** [comment]
Kind of a bizarre mix of circumstances all-around. The underlying issue does seem to be z-fighting caused by simulated terminal bones overlapping with the other visuals. This could potentially be mitigated / reduced by changing the material on the dark blue end bones to use a material that renders it behind the other light blue bones or vice versa so there to mitigate the z-fighting.

I'm not certain why the visuals would update when hovering the laser over the inspector- but overall this seems like a fairly niche issue.

It requires a fairly specific setup to be seen- and only occurs while the debug visuals are enabled.

As mentioned above, this could likely largely be mitigated by a small tweak to the materials used in the debug visuals, but otherwise- the visuals would probably need to be reworked / overhauled because as noted- there were other issues with them as well.

**Comment by ModernBalloonie** [comment]
Yeah it's weird. It isn't just z fighting though, if you check the video, the yellow lines are also flashing and going into different directions. it's like it's delayed.

**Comment by ModernBalloonie** [comment]
Actually, i found another example. Here's another model that does it that a lot of people use. that being The Rikkor: 

[2024-07-09 13-22-29-1.webm](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/82356528/0617be60-bae2-4535-b594-7d1279c15056)

How it looks for one of the frames:
![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/82356528/123bd842-7b4c-4e64-9b5a-e688e65cf3ac)

How it looks for the next frame:
![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/82356528/35f73ab7-a487-432c-98e5-72eae9129323)

**Comment by shiftyscales** [comment]
Same common trend as the other cases though- a bunch of side chains all branching off of the first chain.

**Comment by lxw404** [comment]
I may have some insight on what might be going on, just speculating though:

Debug visuals generated by certain components at the moment seem to have a cap on how many actually exist at once per user.  I found this out by generating a bunch of debug axis visuals at one time using the `DebugTangentSpace` component, and after you go over some amount they start to swap position/rotation every frame randomly with the visuals that are over the limit.  The component must be refreshing these debug visuals at certain intervals, but it doesn't seem to delete them, rather it swaps them somehow.

You can test this by looking at the `__DEBUG` slot in the world under your user, then click on one of the debug visuals and you'll see that after generating too many, they start swapping positions/rotations.

**Comment by ModernBalloonie** [comment]
Apologies for the comment on a post that's really old, but I'm really struggling with this issue. It's getting really difficult to work on some things.

<img width="836" height="1196" alt="Image" src="https://github.com/user-attachments/assets/cf8f19ab-e001-4962-96c6-81b9ff425aeb" />

After the splittening, it's a little bit better, but it still strobes and gives me a headache and prevents me from working on some things, and I REALLY don't want to risk my health, or others when I work on avatars that have bone setups like this.

I don't mean to nag, but I just wanted to have people aware that this is still a problem that's been really getting in the way for me putting some of my VRchat bone setups in Resonite.

In the meantime, if someone can make a mod to help with this, that would be incredibly nice and appreciated.

**Comment by noblereign** [comment]
> In the meantime, if someone can make a mod to help with this, that would be incredibly nice and appreciated.

It seems like the main issue with the current dynamic bone visualizer is that it creates and deletes stuff every single frame which makes the debug visuals freak out
I had an idea and threw together a mod that visualizes the dynamic bones more like how the rig and finger components do, by creating a debug visual for the chain once and just updating it with the simulation. I don't notice any flashing at all this way

https://github.com/noblereign/ResoniteBetterDynBoneVisualizer
